File: RobWork/cpp/src/Sensor.cpp
#include "../inc/Sensor.hpp"

/**
 * @brief Default constructor for SimulatedRGBD object.
*/
SimulatedRGBD::SimulatedRGBD(){}

/**
 * @brief Constructor for SimulatedRGBD object.
 * @param sim_cam The simulated camera to use.
 * @param sim_depth The simulated depth scanner to use.
 * @param intrinsics The intrinsics of the camera. If no intrinsics are provided, the default intrinsics are used.
*/
SimulatedRGBD::SimulatedRGBD(const SimulatedCamera &sim_cam, const SimulatedScanner25D &sim_depth, const cv::Mat &intrinsics) 
    : sim_cam( ownedPtr (new SimulatedCamera(sim_cam)) ),
        sim_depth( ownedPtr (new SimulatedScanner25D(sim_depth)) )
{
    this->sim_depth->open();

    if (intrinsics.empty())
    {
        std::cout << "No intrinsics provided. Using default intrinsics." << std::endl;
        this->intrinsics = (cv::Mat_<double>(3, 3) << 430.0, 0.0,   320.0, 
                                                      0.0,   430.0, 240.0, 
                                                      0.0,   0.0,   1.0);
    }
    else
    {
        this->intrinsics = intrinsics;
    }
    // std::cout << "Intrinsics: \n" << this->intrinsics << std::endl;
}


/**
 * @brief Convert RobWork Image to OpenCV Mat
 * @param image Raw pointer to RobWork Image
 * @param cv_image OpenCV Mat to store the converted image
 * @param imgType Type of image to convert to. Use ImageType::RGB or ImageType::BGR
*/
void SimulatedRGBD::convertImageToCV(const Image *image, cv::Mat &cv_image, int imgType)
{
    cv::Mat _image = cv::Mat(image->getHeight(), image->getWidth(), CV_8UC3, (Image*)image->getImageData());
    // cv::Mat imflip, imflip_mat;
    cv::flip(_image, _image, 1);
    if (imgType == ImageType::RGB)
        _image = _image;
    else if (imgType == ImageType::BGR)
        cv::cvtColor(_image, _image, cv::COLOR_BGR2RGB);
    else
    {
        std::cout << "Image type not supported!" << std::endl;
        std::cout << "Use ImageType::RGB or ImageType::BGR. Image type set to BGR" << std::endl;
    }
    _image.copyTo(cv_image);
}


/**
 * @brief Initialize the simulated RGB camera
 * @param fps Frame rate of the simulated camera
*/
void SimulatedRGBD::initCamera(int fps)
{
    this->sim_cam->setFrameRate(fps);
    this->sim_cam->initialize();
    this->sim_cam->start();
}


/**
 * @brief Initialize the simulated depth scanner
 * @param fps Frame rate of the simulated scanner
*/
void SimulatedRGBD::initScanner25D(int fps)
{
    this->sim_depth->setFrameRate(fps);
    this->sim_depth->open();
}


/**
 * @brief Acquire image from simulated RGBD RealSense sensor. This does not get the image, but waits for the image to be ready.
 * @param state RobWork state of the workcell
 * @param info Simulator update info
*/
void SimulatedRGBD::acquireImage(State &state, const Simulator::UpdateInfo &info)
{
    std::cout << "Acquiring image from simulated RGBD Sensor..." << std::endl;
    int cnt = 0;
    std::cout << "Waiting for image to be ready. Iteration: ";
    
    while (!this->sim_cam->isImageReady())
    {
        std::cout << cnt << ", ";
        this->sim_cam->update(info, state);
        cnt++;
    }
    std::cout << "Image ready!" << std::endl;

}


/**
 * @brief Acquire depth image from simulated RGBD RealSense sensor. This does not get the depth image, but waits for the depth image to be ready.
 * @param state RobWork state of the workcell
 * @param info Simulator update info
*/
void SimulatedRGBD::acquireDepth(State &state, const Simulator::UpdateInfo &info)
{
    std::cout << "Acquiring depth image from simulated RGBD Sensor..." << std::endl;
    int cnt = 0;
    std::cout << "Waiting for depth image to be ready. Iteration: ";
    
    while (!this->sim_depth->isScanReady())
    {
        std::cout << cnt << ", ";
        this->sim_depth->update(info, state);
        cnt++;
    }
    std::cout << "Depth image ready!" << std::endl;
}

/**
 * @brief Get image from simulated RGBD RealSense sensor.
 * @param image_out Image in OpenCV format
 * @param imgType Image type. Use ImageType::RGB or ImageType::BGR
*/
void SimulatedRGBD::getImage(cv::Mat &image_out, int imgType)
{
    const Image *image = this->sim_cam->getImage();
    SimulatedRGBD::convertImageToCV(image, image_out, imgType);
}


/**
 * @brief Get point cloud and depth image from simulated RGBD RealSense sensor.
 * @param pc_out Point cloud pointer in Open3D format
 * @param depthImage_out Depth image in OpenCV format
*/
void SimulatedRGBD::getPointCloudAndDepthImage(PointCloudPtr &pc_out, cv::Mat &depthImage_out)//, const cv::Mat &R, const cv::Mat &t);
{
    const rw::geometry::PointCloud::Ptr pc = ownedPtr( new rw::geometry::PointCloud(this->sim_depth->getScan()) );
    std::vector<rw::math::Vector3Df> data = pc->getData();

    // Convert Vector3Df to eigen vector and Ignore points with -10 depth, because that is just the default value (background)
    std::vector<Eigen::Vector3d> data_eigen;
    std::vector<cv::Point3d> data_cv;

    for (size_t i = 0; i < data.size(); i++)
    {
        if (data[i][2] != -10)
            data_eigen.push_back(Eigen::Vector3d(data[i][0], data[i][1], data[i][2]));
        data_cv.push_back(cv::Point3d(data[i][0], data[i][1], data[i][2]));
    } 

    // Convert point cloud to Open3D point cloud
    pc_out = std::shared_ptr<open3d::geometry::PointCloud>(new open3d::geometry::PointCloud(data_eigen));

    // Flip the point cloud
    Eigen::Matrix4d flip_mat;
    flip_mat << -1, 0, 0, 0,
                0, -1, 0, 0,
                0, 0, 1, 0,
                0, 0, 0, 1;

    pc_out->Transform(flip_mat);

    // Create depth image from point cloud
    cv::Mat depthImage = cv::Mat::zeros(480, 640, CV_16U);
    for (int x = 0; x < depthImage.cols; x++)
    {
        for (int y = 0; y < depthImage.rows; y++)
        {
            auto depth = pc->operator()(x, y)[2];
            int value = 0;
            if (depth != -10)
            {
                value = (int16_t) -10000 * depth;
            }
            
            depthImage.at<int16_t>(y, x) = value;
        }
        
    }
    cv::flip(depthImage, depthImage_out, 1);
}


/**
 * @brief Adds gaussian noise to the depth image.
 * @param depth The input depth image that gets noise added to it.
 * @param mean The mean used for noise generation.
 * @param stddev The standard deviation used for noise generation.
 * @param scale The scale used for noise generation.
*/
void SimulatedRGBD::addDepthNoise(cv::Mat &depth, double mean, double stddev, double scale)
{
    cv::Mat noise = cv::Mat::zeros(depth.size(), CV_64F);
    cv::randn(noise, mean, stddev);
    noise = noise * scale;

    for (int i = 0; i < noise.rows; i++)
    {
        for (int j = 0; j < noise.cols; j++)
        {
            // Scale noise and cast to uint16_t
            auto d = depth.at<uint16_t>(i, j);

            // If depth is 0, point is invalid
            if (d == 0)
                continue;

            // Add noise to depth
            int val = d + round(noise.at<double>(i, j));
            if (val < 0)
                val = 0;

            depth.at<uint16_t>(i, j) = val;
        }
    }
}


/**
 * @brief Closes RobWork simulated camera and depth sensor.
*/
void SimulatedRGBD::close()
{
    this->sim_cam->stop();
    this->sim_depth->close();
}


/**
 * @brief Destructor.
*/
SimulatedRGBD::~SimulatedRGBD()
{
    // this->sim_cam->stop();
    // this->sim_depth->close();
}

File: RobWork/python/Sensor.py
from sdurw import *
from sdurws import *
from sdurwsim import *

import numpy as np
import cv2 as cv
import time

from enum import Enum
class ColorEncoding(Enum):
    GRAY = 0
    RGB = 1
    RGBA = 2
    BGR = 3 
    BGRA = 4  

class FrameGrabber():
    def __init__(self, scanner_type, sim_sensor) -> None:
        self.scanner_type = scanner_type
        self.sim_sensor = sim_sensor
        self.sim_sensor.setFrameRate(100)

        if self.scanner_type == "Camera":
            self.sim_sensor.initialize()
            self.sim_sensor.start()
        elif self.scanner_type == "Scanner25D":
            self.sim_sensor.open()
            if not self.sim_sensor.isOpen():
                raise Exception("Scanner25D could not be opened.")
        else:
            print("Scanner type not supported. Please use 'Camera' or 'Scanner25D'.")


    def rw2image(self, image):
        image_out = np.zeros((image.getHeight(), image.getWidth(), 3), dtype=np.uint8)
        for row in range(image.getHeight()):
            for col in range(image.getWidth()):
                image_out[row, col, 0] = image.getPixelValuei(col, row, 0)
                image_out[row, col, 1] = image.getPixelValuei(col, row, 1)
                image_out[row, col, 2] = image.getPixelValuei(col, row, 2)
        return cv.flip(image_out, 1)

    def acquire_data(self, state, info, printing=False):
        self.sim_sensor.acquire()

        cnt = 0
        if self.scanner_type == "Camera":
            if printing:
                print(f"Acquiring new image from simulated camera")
            if not self.sim_sensor.isImageReady():
                if printing:
                    print("Image is not ready yet. Iteration: ")
            while not self.sim_sensor.isImageReady():
                if printing:
                    print(str(cnt), end=", ")
                self.sim_sensor.update(info, state)
                cnt += 1
            if printing:
                print("")
        elif self.scanner_type == "Scanner25D":
            if printing:
                print(f"Acquiring new depth image from simulated camera")
            if not self.sim_sensor.isScanReady ():
                if printing:
                    print("Scan is not ready yet. Iteration: ")
            while not self.sim_sensor.isScanReady ():
                if printing:
                    print(str(cnt), end=", ")
                self.sim_sensor.update(info, state)
                cnt += 1
            if printing:
                print("")

    def get_data(self, ColorFormat=ColorEncoding.RGB, printing=False):
        data = None
        if self.scanner_type == "Camera":
            if printing:
                print("Getting image from RobWork")

            # Get and convert RobWork image to numpy array
            time_start = time.time()
            image = self.sim_sensor.getImage()
            data = self.rw2image(image)
            time_end = time.time()
            if printing:
                print(f"Time to get image: {time_end - time_start} [s]\n")

            if ColorFormat == ColorEncoding.RGB:
                pass # It is RGB by default
            elif ColorFormat == ColorEncoding.GRAY:
                data = cv.cvtColor(data, cv.COLOR_BGR2GRAY)
            elif ColorFormat == ColorEncoding.BGR:
                data = cv.cvtColor(data, cv.COLOR_RGB2BGR)
        elif self.scanner_type == "Scanner25D":
            if printing:
                print("Getting depth image from RobWork")

            # Get and convert RobWork image to numpy array
            time_start = time.time()
            data = self.sim_sensor.getScan()
            time_end = time.time()
            if printing:
                print(f"Time to get scan: {time_end - time_start} [s]\n")


        return data
    
    def release(self):
        self.sim_sensor.stop()

